src/jubjub/mod.rs

```diff
@@ -434,6 +446,257 @@ impl JubjubBls12 {
+                        .collect::<Vec<u8>>()
+                })
+                .collect();
+
+            file.write(&segment_table_bytes).expect("failed to write segment table bytes");
+        }
+
+        n_segments_written += n_segments_in_range;
+    }
+}
+
+// This function can return less than `n_segments` worth of the exp-table if `first_segment +
+// n_segments` exceeds the length of the exp-table file. If `first_segment` exceeds the number of
+// segments in the exp-table file, and empty vector is returned from this function. It is the user's
+// responsibility to check the number of exp-table segments returned from this function.
+pub fn read_exp_table_range(
+    first_segment: usize,
+    n_segments: usize,
+    path: &str,
+) -> Vec<Vec<Vec<edwards::Point<Bls12, PrimeOrder>>>> {
+    let tmp_params = JubjubBls12 {
+        edwards_d: edwards_d(),
+        // We won't use the remaining fields.
+        pedersen_hash_exp_window_size_supplied: 0,
+        pedersen_hash_exp: vec![],
+        montgomery_a: Fr::zero(),
+        montgomery_2a: Fr::zero(),
+        scale: Fr::zero(),
+        pedersen_hash_generators: vec![],
+        pedersen_circuit_generators: vec![],
+        fixed_base_generators: vec![],
+        fixed_base_circuit_generators: vec![],
+        exp_table_path: None,
+    };
+
+    let mut file = File::open(path).expect("failed to open Pedersen Hash params file");
+    // We can skip the number of segments in this file (the first 4 bytes), we don't use that value
+    // here.
+    let mut offset = 4u64;
+
+    let window_size = {
+        file.seek(SeekFrom::Start(offset)).unwrap();
+        let mut buf = [0u8; 4];
+        file.read_exact(&mut buf).expect("failed to read bytes 4-8 from the exp-table file");
+        offset += 4;
+        u32::from_le_bytes(buf)
+    };
+
+    // Calculate the number of bytes we have to seek forward per segment in the exp-table file.
+    let n_windows_per_segment = {
+        let n_windows_per_segment = fs::Fs::NUM_BITS as f32 / window_size as f32;
+        if n_windows_per_segment.fract() == 0.0 {
+            (n_windows_per_segment as usize) + 1
+        } else {
+            n_windows_per_segment.ceil() as usize
+        }
+    };
+    let n_points_per_window = 2usize.pow(window_size);
+    let n_bytes_per_point = size_of::<Fr>();
+    let n_bytes_per_segment = n_windows_per_segment * n_points_per_window * n_bytes_per_point;
+
+    // Move the offset to the first byte of the first segment that we want to read.
+    offset += n_bytes_per_segment as u64 * first_segment as u64;
+    file.seek(SeekFrom::Start(offset)).unwrap();
+
+    // Read in the desired range of the exp-table.
+    let mut exp_table = vec![];
+
+    for _ in 0..n_segments {
+        let mut segment_windows = vec![];
+        for _ in 0..n_windows_per_segment {
+            let mut window = vec![];
+            for _ in 0..n_points_per_window {
+                // TODO: reduce the number of file reads.
```

Comment:
just wrap the file into a `BufReader`, that way you can keep the reads as you do them right now, but they are more efficent under the hood

Reply:
Good suggestion, I didn't do that because you can't do absolute seeking in buffered I/O. But I found a way to make it work using `BufReader`.